doc/developer/design/20210601_build_mirrelationexpr.md

```diff
@@ -0,0 +1,341 @@
+    generate-series-int32
+    [#1]
+  )
+  [(call-binary eq #0 #3)]
+)
+```
+
+but implementing the syntax itself is onerous because Rust can't take a string
+and just convert it to an enum. The code for constructing an `BinaryFunc`
+would look like:
+
+```
+match func {
+"eq" => BinaryFunc::Eq
+"isregexpmatch" => {
+    let case_insensitive = self.next_sexp().parse::<bool>();
+    BinaryFunc::IsRegexpMatch {case_insensitive}
+}
+<repeat for every single BinaryFunc variant>
+}
+```
+
+As such, currently Justin's system only supports the following flavors:
+* Of `MirScalarExpr` => columns, numeric literals, bool literals
+* Of `MirRelationExpr` => get, let, map, filter, project, constant, join (only
+`JoinImplementation::Unimplemented` ), union, negate, arrange-by
+
+Figuring out how to more efficiently implement the syntax has been an ongoing
+issue. There has been some earlier design discussions on this, which are
+documented in the comments of [#5684].
+
+[#5684]: https://github.com/MaterializeInc/materialize/issues/5684
+
+## Goals
+
+* Allow the easy construction of all, or at
+  least the vast majority of, `Mir{Relation, Scalar}Expr`. Being able to do this
+  will help a lot with:
+  * Optimization/transform unit testing. 
+  * Testing the dataflow layer directly. Currently, there are some branches
+    that exist so Materialize does not crash if there are optimizer bugs. These
+    branches end up untested because we have eliminated all known optimizer
+    bugs that would trigger those branches.
+  * Experimentation. Being able to construct arbitrary `Mir`s can allow for
+    testing the effects of proposed transforms without having to write code for
+    the transform. 
+  * Debugging. Often times you want to find a simpler reproduction for a
+    complicated query that is failing, but changing the SQL may result in a
+    complete different `Mir` representation.
+* Create a system that can easily ported to supporting the easy construction
+  of other complex internal expressions so that we don't have to go through
+  this again if we want to test or play around with other stuff. Example other
+  internal expressions:
+  * `LIRR`
+  * future `Lir` expressions
+  * possibly `Hir` expressions? We have been talking about doing some rewrites before the decorrelation stage.
+  * `DataflowDescs` to test optimization across dataflows?
+
+## Non-Goals
+
+* The consensus from the earlier discussion is that the
+  `Mir{Relation, Scalar}Expr` builder should not bother with attempting to
+  infer the type of a column. E.g. when doing a sum reduce, it is up to the
+  test-writer to select the correct `AggregateFunc` from `SumInt32`,
+  `SumInt64`, etc.
+
+## Description
+
+### General Approach
+
+Let's call this approach "Minimal boilerplate".
+
+The goal is to convert Justin's syntax into a JSON string such that
+deserializing the JSON using [`serde_json`](https://docs.serde.rs/serde_json/)
+will produce your desired enum.
+
```

Comment:
I read this a few times, but I think this paragraph is the actual proposal. If that is correct, can we change the text to be more concrete, active-voice-y. Like:

> The concrete proposal is to retain Justin's syntax, and invest in a mechanism that can automatically convert it in to a JSON formatted string that can be correctly deserialized by `serde_json`.

I think it's pretty much the same minus the "The goal is" which idk sounds like it is part of a problem statement rather than a solution.

If I haven't totally misunderstood, there are some similar language issues in the next paragraphs (e.g. "To get from ... to ... requires ..." -> "We can translate ... into ... by way of ..."; "the specification ... can be automatically converted" -> "We can convert the specification ... automatically". I wasn't sure reading this if you actually thought it was possible, but if you are / want to propose it .. I was confused. :D

Reply:
Will update proposal with more active language.